evm/vm/base.py

```diff
@@ -147,6 +189,131 @@ def mine_block(self, *args, **kwargs):
 
         return block
 
+    def pack_block(self, block, *args, **kwargs):
+        """
+        Pack block for mining.
+
+        :param bytes coinbase: 20-byte public address to receive block reward
+        :param bytes uncles_hash: 32 bytes
+        :param bytes state_root: 32 bytes
+        :param bytes transaction_root: 32 bytes
+        :param bytes receipt_root: 32 bytes
+        :param int bloom:
+        :param int gas_used:
+        :param bytes extra_data: 32 bytes
+        :param bytes mix_hash: 32 bytes
+        :param bytes nonce: 8 bytes
+        """
+        if 'uncles' in kwargs:
+            block.uncles = kwargs.pop('uncles')
+            kwargs.setdefault('uncles_hash', keccak(rlp.encode(block.uncles)))
+
+        header = block.header
+        provided_fields = set(kwargs.keys())
+        known_fields = set(tuple(zip(*BlockHeader.fields))[0])
+        unknown_fields = provided_fields.difference(known_fields)
+
+        if unknown_fields:
+            raise AttributeError(
+                "Unable to set the field(s) {0} on the `BlockHeader` class. "
+                "Received the following unexpected fields: {0}.".format(
+                    ", ".join(known_fields),
+                    ", ".join(unknown_fields),
+                )
+            )
+
+        for key, value in kwargs.items():
+            setattr(header, key, value)
+
+        # Perform validation
+        self.state.validate_block(block)
+
+        return block
+
+    @classmethod
+    def finalize_block(cls, vm_state, block):
+        """
+        Finalize the given block (set rewards).
+        """
+        block_reward = cls.get_block_reward(block.number) + (
+            len(block.uncles) * cls.get_nephew_reward(block.number)
+        )
+
+        with vm_state.state_db() as state_db:
+            state_db.delta_balance(block.header.coinbase, block_reward)
+            vm_state.logger.debug(
+                "BLOCK REWARD: %s -> %s",
+                block_reward,
+                block.header.coinbase,
+            )
+
+            for uncle in block.uncles:
+                uncle_reward = cls.get_uncle_reward(block.number, uncle)
+                state_db.delta_balance(uncle.coinbase, uncle_reward)
+                vm_state.logger.debug(
+                    "UNCLE REWARD REWARD: %s -> %s",
+                    uncle_reward,
+                    uncle.coinbase,
+                )
+        block.state_root = vm_state.block_header.state_root
+
+        return block, vm_state
+
+    @classmethod
+    def create_block(
+            cls,
+            transactions,
+            transaction_witnesses,
```

Comment:
not sure if this is an improvement but...

If I understand correctly, we require that `len(transaction) == len(transaction_witnesses)`.  Rather than passing these in as separate lists, it might be good to use a data structure like the following:

```
transactions_and_witnesses = (
    (transaction_0, transaction_witness_0),
    (transaction_1, transaction_witness_1),
    (transaction_2, transaction_witness_2),
    ...
)
```

This approach builds in some minimal validation since you can then loop over them like this:

```python
for index, (transaction, transaction_witness) in enumerate(transactions_and_witnesses):
    ...
```

Not sure this is a huge improvement, but since the two arguments are inherently linked together, it might be nice to accept them already linked.

Reply:
Yes, we do require that `len(transaction) == len(transaction_witnesses)`. 
From sharding doc:
```python
Transaction package format
    [
        [nonce, acct, data....],    # transaction body (see below for specification)
        [node1, node2, node3....]   # witness
    ]
```
I will modify it to the form you suggested, thank you. :)